src/Microsoft.Framework.Logging/LoggerExtensions.cs

```diff
@@ -18,6 +18,131 @@ public static class LoggerExtensions
         private static readonly Func<object, Exception, string> _loggerStructureFormatter = (state, error)
             => LoggerStructureFormatter((ILoggerStructure)state, error);
 
+        private static readonly Func<object, Exception, bool, string> SecureMessage =
+            (messages, error, secure) =>
+            {
+                if (secure)
+                {
+                    return ((string[])messages)[0];
+                }
+                return ((string[])messages)[1];
+            };
+
+        private static readonly Func<object, Exception, bool, string> SecureMessageAndError =
+            (messages, error, secure) =>
+            {
+                if (secure)
+                {
+                    return string.Format(CultureInfo.CurrentCulture, "{0}{1}{2}", ((string[])messages)[0], Environment.NewLine, error);
+                }
+                return string.Format(CultureInfo.CurrentCulture, "{0}{1}{2}", ((string[])messages)[1], Environment.NewLine, error);
+            };
+
+        private static readonly Func<object, Exception, bool, string> _secureLoggerStructureFormatter =
+            (state, error, secure) =>
+            {
+                var formatter = (ILoggerStructure[]) state;
+
+                if (secure)
+                {
+                    return LoggerStructureFormatter(formatter[0], error);
+                }
+                return LoggerStructureFormatter(formatter[1], error);
+                
+            };
+
+        //------------------------------------------DEBUG------------------------------------------//
+
+        /// <summary>
+        /// Writes a verbose log message.
+        /// </summary>
+        /// <param name="logger">The <see cref="ILogger"/> to write to.</param>
+        /// <param name="data">The message to log at verbose level.</param>
+        /// <param name="secureData">The message to log at debug level.</param>
+        // FYI, this field is called data because naming it message triggers CA1303 and CA2204 for callers.
+        public static void WriteDebug([NotNull] this ILogger logger, string secureData, string data)
```

Comment:
Not really a fan of calling I "secureXyz" in various places. This is the only place that the security nomenclature seems to be used. Everywhere else it's a distinction between "debug" and not, and that seems orthogonal to the distinction between "secure" and not.